File: RogueTraderSystemGenerator/Nodes/PirateShipsNode.cs
using System;
using System.Globalization;
using System.Runtime.Serialization;
using System.Windows.Documents;

namespace RogueTraderSystemGenerator.Nodes
{
    [DataContract]
    class PirateShipsNode : NodeBase
    {
        [DataMember]
        private bool _pirateDenContainsWayfarerStation;
        [DataMember]
        private Species _pirateSpecies;

        public PirateShipsNode() 
        {
            _nodeName = "Pirate Den";
        }

        public override void ResetVariables()
        {
            base.ResetVariables();
            _pirateDenContainsWayfarerStation = false;
            _pirateSpecies = Species.None;
        }

        public override void GenerateFlowDocument()
        {
            _flowDocument = new FlowDocument();
            DocBuilder.AddHeader(ref _flowDocument, NodeName, 2);
            if (!String.IsNullOrEmpty(Description))
                DocBuilder.AddContentLine(ref _flowDocument, "Description", new DocContentItem(Description));
            if (_pirateDenContainsWayfarerStation)
                DocBuilder.AddContentLine(ref _flowDocument, "Pirate Den", new DocContentItem("The pirates in this system are based around a space station (such as a Wayfarer Station).", 210, "Wayfarer Station", RuleBook.CoreRuleBook));
            if (Children.Count > 0)
                DocBuilder.AddContentLine(ref _flowDocument, "Number of Pirate Ships Present", new DocContentItem(Children.Count.ToString(CultureInfo.InvariantCulture)));
        }

        public override void Generate()
        {
            if (Globals.RollD10() >= 5)
                _pirateDenContainsWayfarerStation = true;

            _pirateSpecies = Species.Random;
            while (true)
            {
                _pirateSpecies = StarshipTools.GetRandomSpecies();
                if (_pirateSpecies == Species.Human ||
                    (_pirateSpecies == Species.Ork && Properties.Settings.Default.BookBattlefleetKoronus) ||
                    (_pirateSpecies == Species.Eldar && Properties.Settings.Default.BookBattlefleetKoronus) ||
                    (_pirateSpecies == Species.RakGol && Properties.Settings.Default.BookBattlefleetKoronus) ||
                    (_pirateSpecies == Species.ChaosReaver && Properties.Settings.Default.BookBattlefleetKoronus) ||
                    (_pirateSpecies == Species.DarkEldar && Properties.Settings.Default.BookBattlefleetKoronus && Properties.Settings.Default.BookTheSoulReaver))
                {
                    break;
                }
            }

            //_pirateDenShips = new List<DocContentItem>();
            int tempValue1 = Globals.RollD5() + 4;
            int tempValue2 = Globals.RollD5() + 4;
            int totalShips = tempValue1 < tempValue2 ? tempValue1 : tempValue2;
            for (int i = 0; i < totalShips; i++)
            {
                Starship ship = StarshipTools.GetRandomPirateShip(_pirateSpecies);
                ShipNode node = new ShipNode(ship) {Parent = this};
                Children.Add(node);
            }
        }

        public void AddNewShip()
        {
            Starship ship = StarshipTools.GetRandomPirateShip(_pirateSpecies);
            ShipNode node = new ShipNode(ship) {Parent = this};
            Children.Add(node);
        }
    }
}

File: electron-app/js/nodes/pirateShipsNode.js
// PirateShipsNode.js
class PirateShipsNode extends NodeBase {
    constructor(id = null) {
        super(NodeTypes.PirateShips, id);
        this.nodeName = 'Pirate Den';
        this.fontStyle = 'italic';
        this.fontForeground = '#e74c3c';
        this.headerLevel = 3; // H3: Feature node
        this.containsWayfarerStation = false; // parity: _pirateDenContainsWayfarerStation
        this.pirateSpecies = Species.None;    // parity: _pirateSpecies (kept internal; not surfaced in description)
        this.numShips = 0;                    // cached count for description only
    }

    generate() {
        super.generate();
        
        // Station presence (d10 >=5)
        this.containsWayfarerStation = (RollD10() >= 5);

        // Choose species similar to WPF gating
        const enabled = window.APP_STATE.settings.enabledBooks || {};
        let safety = 200; // guard against infinite loop on misconfigured data
        do {
            this.pirateSpecies = window.StarshipToolsData.getRandomSpecies();
            if (this.pirateSpecies === Species.Human) break;
            if (this.pirateSpecies === Species.Ork && enabled.BattlefleetKoronus) break;
            if (this.pirateSpecies === Species.Eldar && enabled.BattlefleetKoronus) break;
            if (this.pirateSpecies === Species.RakGol && enabled.BattlefleetKoronus) break;
            if (this.pirateSpecies === Species.ChaosReaver && enabled.BattlefleetKoronus) break;
            if (this.pirateSpecies === Species.DarkEldar && enabled.BattlefleetKoronus && enabled.TheSoulReaver) break;
            safety--;
        } while (safety > 0);

        // Number of ships: 2d5 keep lowest-ish (WPF: RollD5()+4 twice, take min)
        const temp1 = RollD5() + 4;
        const temp2 = RollD5() + 4;
        this.numShips = Math.min(temp1, temp2);
        // NOTE(Parity Confirmed): WPF also uses two (RollD5()+4) rolls and takes the minimum; no other feature modifies ship count.

        // Create ships as children
        this.children = [];
        for (let i = 0; i < this.numShips; i++) this._addNewShipInternal();
        this.updateDescription();
    }

    _addNewShipInternal() {
        const shipData = window.StarshipToolsData.getRandomPirateShip(this.pirateSpecies);
        const shipNode = new ShipNode();
        shipNode.setShip(shipData);
        this.addChild(shipNode);
    }

    // Parity public helper (mirrors C# AddNewShip)
    addNewShip() {
        this._addNewShipInternal();
        this.numShips = this.children.length;
        this.updateDescription();
    }

    updateDescription() {
        // No duplicate header - base class will add H3
        // NOTE (Parity): Wayfarer Station line should include page 210 (Core Rulebook) when page refs enabled.
        const showPages = window.APP_STATE?.settings?.showPageNumbers;
        const wayfarerRef = showPages ? ` <span class="page-reference">${createPageReference(210,'Wayfarer Station')}</span>` : '';
        let desc = '';
        if (this.containsWayfarerStation) {
            desc += `<p>The pirates in this system are based around a space station (such as a Wayfarer Station).${wayfarerRef}</p>`;
        }
        // Match C# wording: label style "Number of Pirate Ships Present"
        desc += `<p><strong>Number of Pirate Ships Present:</strong> ${this.numShips}</p>`;
        this.description = desc;
    }

    getContextMenuItems() {
        return [
            { label: 'Add Pirate Ship', action: 'add-pirate-ship' },
            { type: 'separator' },
            { label: 'Edit Description', action: 'edit-description' }
        ];
    }

    toJSON() {
        const json = super.toJSON();
        json.containsWayfarerStation = this.containsWayfarerStation;
        json.pirateSpecies = this.pirateSpecies;
        json.numShips = this.numShips;
        return json;
    }

    static fromJSON(data) {
        const node = new PirateShipsNode(data.id);
        
        // Restore base properties
        Object.assign(node, {
            nodeName: data.nodeName || 'Pirate Den',
            description: data.description || '',
            customDescription: data.customDescription || '',
            pageReference: data.pageReference || '',
            isGenerated: data.isGenerated || false,
            fontWeight: data.fontWeight || 'normal',
            fontStyle: data.fontStyle || 'normal',
            fontForeground: data.fontForeground || '#e74c3c'
        });
        
        // Restore pirate ships-specific properties
        node.containsWayfarerStation = data.containsWayfarerStation || false;
        node.pirateSpecies = data.pirateSpecies || Species.None;
        node.numShips = data.numShips || 0;
        
        // Restore children
        if (data.children) {
            for (const childData of data.children) {
                const restoredChild = window.restoreChildNode(childData);
                node.addChild(restoredChild);
            }
        }
        
        // Update numShips to match actual children count
        node.numShips = node.children.length;
        
        // Regenerate description if not saved
        if (!node.description) {
            node.updateDescription();
        }
        
        return node;
    }
}

window.PirateShipsNode = PirateShipsNode;